keps/sig-storage/20190120-ExecutionHook-apichange.md

```diff
@@ -227,6 +227,91 @@ by the snapshot controller and the default is `false`.
 The user can use Annotations to define the execution hook if it is not in the
 container struct.
 
+We also considered the following alternatives based on review feedbacks.
+
+### Alternative Option 1
+
+Keep `ExecutionHook` the same as what is defined in the proposal section. Add `ExecutionHook`
+to `Lifecycle`.
+
+```
+type Lifecycle struct {
+        // PostStart is called immediately after a container is created.  If the handler fails, the container
+        // is terminated and restarted.
+        // +optional
+        PostStart *Handler
+        // Defines the hooks to trigger an action in a container
+        // +optional
+        ExecutionHooks []ExecutionHook
+        // PreStop is called immediately before a container is terminated.  The reason for termination is
+        // passed to the handler.  Regardless of the outcome of the handler, the container is eventually terminated.
+        // +optional
+        PreStop *Handler
+}
+```
+
+### Alternative Option 2
+
+Add `Name`, `Type`, and `TimeOutSeconds` from the `ExecutionsHook` struct directly to the Handler struct. In that case we have to make `Type` optional for backward compatibility.
+
+```
+type Handler struct {
+        // One and only one of the following should be specified.
+        // Exec specifies the action to take.
+        // +optional
+        Exec *ExecAction
+        // HTTPGet specifies the http request to perform.
+        // +optional
+        HTTPGet *HTTPGetAction
+        // TCPSocket specifies an action involving a TCP port.
+        // TODO: implement a realistic TCP lifecycle hook
+        // +optional
+        TCPSocket *TCPSocketAction
+        // Name of an execution hook
+        // +optional
+        Name string
+        // Type of an execution hook
+        // +optional
+        Type string
```

Comment:
Type doesn't seem necessary. Will a given item need multiple freeze hooks and multiple thaw hooks?